frontend/src/Frontend/UI/Widgets.hs

```diff
@@ -176,59 +177,102 @@ uiRealInputElement cfg = do
       & initialAttributes %~ addInputElementCls . addNoAutofillAttrs
         . (<> ("type" =: "number" <> "step" =: stepSize <> "min" =: stepSize))
-      & inputElementConfig_setValue %~ (\e -> leftmost [fmapMaybe parseAndRound e, rounded])
-    widgetHold_ blank (bool blank (elClass "span" "real-input__rounded" (text $ "Rounded to " <> (tshow prec) <> " places")) <$> eHasRounded)
-    let parsedValue = fmap fst . parseDecimal <$> value r
-        parsedInput = fmapMaybe parseDecimal (_inputElement_input r)
-        -- Trim the users input if we had to round it
-        rounded = fmap (showDecimal . fst) $ ffilter snd parsedInput
-        eHasRounded = fmap snd parsedInput
-  pure
-    ( r
-    , (fmap fromDecimal <$> parsedValue
-      , fromDecimal . fst <$> parsedInput
-      )
-    )
-  where
-    showDecimal :: Decimal -> Text
-    showDecimal = tshow
-
-    parseAndRound :: Text -> Maybe Text
-    parseAndRound t = showDecimal . fst <$> parseDecimal t
-
-    -- Returns the decimal and whether or not it needed rounding to 'prec'
-    parseDecimal :: Text -> Maybe (Decimal, Bool)
-    parseDecimal t = ffor (readMay $ T.unpack t) $ \decimal ->
-      (D.roundTo prec decimal, D.decimalPlaces decimal > prec)
+    widgetHold_ blank $ ffor (fmap snd input) $ traverse_ $
+      elClass "span" "real-input__rounded" . text
+  pure (ie, (fmap . fmap) fromDecimal val, fmap (fromDecimal . fst) input)
 
+  where
     stepSize = "0." <> T.replicate (fromIntegral prec - 1) "0" <> "1"
+    parse = tread
+
+    sanitize :: Decimal -> Maybe (Decimal, Text)
+    sanitize decimal = asum
+      [ (0, "Cannot be negative")
+        <$ guard (decimal < 0)
+      , (D.roundTo prec decimal, ("Rounded to " <> tshow prec <> " places"))
+        <$ guard (D.decimalPlaces decimal > prec)
+      ]
 
+-- TODO: correct floating point decimals
 uiIntInputElement
   :: DomBuilder t m
-  => InputElementConfig er t (DomBuilderSpace m)
-  -> m (InputElement er (DomBuilderSpace m) t)
-uiIntInputElement cfg = do
-    r <- inputElement $ cfg & initialAttributes %~
-            (<> ("type" =: "number")) . addInputElementCls . addNoAutofillAttrs
-    pure $ r
-      { _inputElement_value = fmap fixNum $ _inputElement_value r
-      , _inputElement_input = fmap fixNum $ _inputElement_input r
-      }
+  => MonadFix m
+  => Maybe Integer
+  -> Maybe Integer
+  -> InputElementConfig er t (DomBuilderSpace m)
+  -> m (InputElement er (DomBuilderSpace m) t, Event t Integer)
+uiIntInputElement mmin mmax cfg = do
+    (r, _, input) <- uiCorrectingInputElement (tread . fixNum) sanitize tshow $ cfg & initialAttributes %~
+      ((<> numberAttrs) . addInputElementCls . addNoAutofillAttrs)
+
+    pure (r
+          { _inputElement_value = fmap fixNum $ _inputElement_value r
+          , _inputElement_input = fmap fixNum $ _inputElement_input r
+          }
+         , fst <$> input)
+
   where
     fixNum = T.takeWhile (/='.')
+    numberAttrs = mconcat $ fmapMaybe id
```

Comment:
`min`/`max` attributes only seem to be applied to arrow-key updates? I could certainly add a minus sign or just add digits to the number without them affecting anything

Reply:
Yeah, I think so. That's what I'd expect. It's not really validation.